iree/compiler/Dialect/HAL/Transforms/RepeatDispatches.cpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALTypes.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+namespace {
+
+static llvm::cl::opt<int> repeatDispatchNum{
+    "iree-hal-repeat-dispatch-num",
+    llvm::cl::desc("The number of times to repeat dispatches."),
+    llvm::cl::init(1),
+};
+
+// A pass converting the IREE flow dialect into the IREE HAL dialect.
+class RepeatDispatchesPass
+    : public PassWrapper<RepeatDispatchesPass, OperationPass<FuncOp>> {
+ public:
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<HALDialect, StandardOpsDialect, scf::SCFDialect>();
+  }
+
+  void runOnOperation() override {
+    if (repeatDispatchNum == 1) return;
+    FuncOp f = getOperation();
+    f.walk([&](HAL::CommandBufferDispatchOp op) {
+      OpBuilder builder(op);
+      Location loc = op.getLoc();
+      Value zero = builder.create<ConstantIndexOp>(loc, 0);
+      Value one = builder.create<ConstantIndexOp>(loc, 1);
+      Value ub = builder.create<ConstantIndexOp>(loc, repeatDispatchNum);
+      auto forOp = builder.create<scf::ForOp>(
```

Comment:
Just unroll here yourself instead of using the for loop at runtime - the runtime loop would only be useful if you were taking the count as an argument to the invocation.

That's a possible improvement: you could define a hal.variable of the count, create a new export function to set it (`_set_benchmark_batch_size(index)`), and then use that as the range here. Then you have control when executing instead of needing to recompile to change the batch count.

If just doing it statically, though, then avoid the runtime loop.

Reply:
The possible improvement sounds good to me. But it is not a trivial change to me.

It looks like I have to create the `count` hal.variable during the pass, and query the variable (by `hal.variable.load`?) to set the loop boundary? 

I don't know how to set it by adding `_set_benchmark_batch_size` during runtime. How do I know the variable name or can we make it fixed? If we can make it fixed, how do we write data to the variable (or should it be a buffer?)?

Reply:
yep - you'd add the variable, then add the function that just performs a hal.variable.store of the parameter to that variable so literally 2 ops: hal.variable.storage + return:
```mlir
hal.variable @_dispatch_batch_count mutable : index = 1
func @_set_benchmark_batch_size(%new_size : index) {
  hal.variable.store %new_size, @_dispatch_batch_count : index
  return
}
```

then you use iree_vm_invoke to call that function with your integer argument. I do things like that in the tflite bindings, for example, as they have some stateful getters/setters:
https://github.com/google/iree/blob/1967f658e318197b3cb4c201870b0ec72ad3b170/bindings/tflite/interpreter.c#L444-L461

```c
iree_vm_value_t value = iree_vm_value_make_i32((int32_t)new_size);
iree_vm_list_set_value(inputs, 0, &value);
status = iree_vm_invoke(context, set_benchmark_batch_size_fn,
                        /*policy=*/NULL, inputs, /*outputs=*/NULL,
                        allocator);
```

and wherever you need the value in the program you can just `hal.variable.load`:
```mlir
%batch_size = hal.variable.load @_dispatch_batch_count : index
scf.for %iv = %c0 to %batch_size step %c1 {
  hal.command_buffer.dispatch ....
}
```

A follow-on for sure, but would remove all the required compile-time batch count stuff here and you could always insert it when benchmarking is enabled - if the benchmark tool never changes the value and it's left at the default of 1 then that's fine.

Leaving this here for future reference of whoever does that improvement :)

Reply:
This is really awesome, and we definitely want it. I'd love to spend some time on this! Thanks for the detailed explanation.